Add 'No selection' option to GID selector

carma12 opened the pull request:
To replicate the same behavior as in the current WebUI, the 'GID' field located in the 'Add user' modal should have a 'No selection' option. Also, a new GID value can be added and associated to a newly-created user.

Fixes: https://github.com/freeipa/freeipa-webui/issues/236

pvoborni commented:
Note: in the review which lead to #236 was mentioned: `In user add dialog, I cannot unset GID when I selected some, also the GID field should allow to enter arbitrary integer.` 

This and #236 is about the first part. But it won't solve the other: "the GID field should allow to enter arbitrary integer" - see the old UI.

carma12 commented:
> Note: in the review which lead to #236 was mentioned: `In user add dialog, I cannot unset GID when I selected some, also the GID field should allow to enter arbitrary integer.`
> 
> This and #236 is about the first part. But it won't solve the other: "the GID field should allow to enter arbitrary integer" - see the old UI.

I think the [PF Select component](https://www.patternfly.org/components/forms/form-select) currently doesn't allow any writing functionality, but I can check with them in case there is some tweak/alternative component we can use for that. 

If not, I can open a discussion with Brian to find out how to approach this design-wise. I will set the 'WIP' tag on this PR while discussing this topic.

carma12 commented:
I checked with the PatternFly team and there is a type of Selector that allows [type-ahead](https://www.patternfly.org/components/menus/select/#typeahead) functionality. So I will replace the existing Select component with this one and modify the description of the PR.

carma12 commented:
@pvoborni - I have added a [new commit](https://github.com/freeipa/freeipa-webui/pull/254/commits/1aa3e0fb3162f7bd2ff5bc30fcc0b06885c75e14) to handle the type-ahead functionality in the GID field. Feel free to review it and share your thoughts on it. Thanks!

pvoborni commented on src/components/modals/AddUser.tsx:
```diff
@@ -286,38 +287,36 @@ const AddUser = (props: PropsToAddUser) => {
-  };
+  const [GIDs, setGIDs] = useState<GroupId[]>([]);
+  const [gidSelected, setGidSelected] = useState<string>("");
+  const [gidOptions, setGidOptions] = React.useState<SelectOptionProps[]>(
```
This state seems unnecessary to me. It can replaced with:

```ts
const gitOptions = getGitOptions(GIDs); // new name based on other comment
```

With that the 
```ts
React.useEffect(() => {
    setGidOptions(assignGidOptions());
  }, [GIDs]);
```
is not needed. And the component will avoid one render on setting the state.

pvoborni commented on src/components/TypeAheadSelectWithCreate.tsx:
```diff
@@ -0,0 +1,263 @@
+  onChangeSelectOptions: (selectOptions: SelectOptionProps[]) => void;
+  selected: string;
+  onChangeSelected: (selected: string) => void;
+}
```
The interface is not very similar to other, e.g. PatternFly components.  It's also unusual to to have props of other component as input to this component (i.e. the SelectOptionProps[]). 

```ts

interface TypeAheadOption {
  value: String;
  label: React.ReactNode; // which could be a string, as well
}

interface PropsToTypeAheadSelectWithCreate {
  options: TypeAheadOption[];
  onOptionsChange: (options: TypeAheadOption[]) => void; // is this needed?
  selected: String;
  onSelectedChange: (selected: string) => void;
}
```

carma12 replied:
Changing that in the code.

pvoborni replied:
OK, let's keep it. 

But exposing internal implementation to external API might sometime lead to unexpected when somebody decides to use it in unexpected/creative way.

pvoborni commented on src/components/TypeAheadSelectWithCreate.tsx:
```diff
@@ -0,0 +1,263 @@
+
+interface PropsToTypeAheadSelectWithCreate {
+  initialSelectOptions: SelectOptionProps[];
+  onChangeSelectOptions: (selectOptions: SelectOptionProps[]) => void;
```
Is this useful? I see that it is used in `AddUser` but is there a reason for it?

carma12 replied:
This is meant to update the selection list options that are retrieved from the `AddUser`, IIRC. As this bit was included in the [Select with Type ahead](https://www.patternfly.org/components/menus/select/#multiple-typeahead-with-create-option) original code, I decided to include it as well as part of the functionality.

pvoborni replied:
The fact that it was in the example code is not a reason to have it here. But in this case it was not even in the example. The example implements a component `SelectTypeaheadCreatable` that has no externally definable props. I.e. these functions were internal to the component.  

The only reason to expose them would be to allow the parent component to work with the new states. E.g. to use them in adding new users. But this imho not needed.

Also with 
```ts
let gidOptions = getGidOptions(GIDs);

  const onChangeGidOptions = (newList: SelectOptionProps[]) => {
    gidOptions = newList;
  };
```
it has no effect now.

pvoborni reviewed (commented):
I've provided some simplifying/readability suggestions. 

But I'd do one more, I'd say more important. The `AddUser` modal is "polluted" with a "large" GID logic that makes the `AddUser` component hard to read. It would be more readable to move out the GID logic to a separate component e.g. called `GidSelector` which would wrap `TypeAheadSelectWithCreate` the same way as the dialog does. Thus the dialog could have only:

```ts
import GIdSelector from "../form/GIdSelector"

const [gid, setGid] = useState("");

<GidSelector selected={gid} onSelectedChange={setGid} />
```

carma12 commented on src/components/TypeAheadSelectWithCreate.tsx:
```diff
@@ -0,0 +1,263 @@
+              "-"
+            )}`}
+            {...option}
+            ref={null}
```
This was already included in the PF component code for [Type ahead](https://www.patternfly.org/components/menus/select/#multiple-typeahead-with-create-option) (check line 216). 

I'm not sure why they decided to add that, though. I dare to guess that it is because of cleanup reasons or good practices. Check this [link](https://stackoverflow.com/questions/72210368/react-ref-null-on-conditional-render). In this case, removing it doesn't seem to affect the functionality of the current code.

carma12 commented:
@pvoborni - As suggested, I removed the `onOptionsChange` prop from the `TypeAheadSelectWithCreate` component and adjusted the `AddUser` one. You were right, it is not needed and works fine this way. Feel free to review it.

pvoborni reviewed (approved):
LGTM
